kind.go

```diff
@@ -45,48 +47,204 @@ func (m Maturity) String() string {
+	// loaded and validated kind definition.
 	Def() Def[CoreProperties]
+
+	// ToBytes takes a []byte and a decoder, validates it against schema, and
+	// if validation is successful, unmarshals it into an UnstructuredResource.
+	// ToBytes(UnstructuredResource, codec Encoder) ([]byte, error)
 }
 
+// Custom is the dynamically typed runtime representation of a Grafana custom kind
+// definition. It is one in a family of interfaces, see [Kind] for context.
+//
+// A Custom kind provides interactions with its corresponding [Resource] using
+// [UnstructuredResource].
+//
+// Custom kinds are declared in Grafana extensions, rather than in Grafana core. It
+// is likely that this distinction will go away in the future, leaving only
+// Custom kinds.
 type Custom interface {
-	Kind
+	ResourceKind
 
-	// TODO docs
+	// Def returns a wrapper around the underlying CUE value that represents the
+	// loaded and validated kind definition.
 	Def() Def[CustomProperties]
 }
 
+// Composable is the untyped runtime representation of a Grafana core kind definition.
+// It is one in a family of interfaces, see [Kind] for context.
+//
+// TODO sort out the Go type used for generic associated...objects? do we even need one?
 type Composable interface {
 	Kind
 
-	// TODO docs
+	// Def returns a wrapper around the underlying CUE value that represents the
+	// loaded and validated kind definition.
 	Def() Def[ComposableProperties]
 }
+
+// TypedCore is the statically typed runtime representation of a Grafana core
+// kind definition. It is one in a family of interfaces, see [Kind] for context.
+//
+// A TypedCore provides typed interactions with the [Resource] type given as its
+// generic type parameter. As it embeds [Core], untyped interaction is also available.
+//
+// A TypedCore is created by calling [BindCoreResource] on a [Core] with a
+// Go type to which it is assignable (see [thema.BindType]).
+type TypedCore[R Resource] interface {
+	Core
+
+	// TypeFromBytes is the same as [Core.FromBytes], but returns an instance of the
+	// associated generic struct type instead of an [UnstructuredResource].
+	TypeFromBytes(b []byte, codec Decoder) (R, error)
+}
+
+// TypedCustom is the statically typed runtime representation of a Grafana core kind definition.
+// It is one in a family of interfaces, see [Kind] for context.
+//
+// A TypedCustom provides typed interactions with the [Resource] type given as its
+// generic type parameter. As it embeds [Custom], untyped interaction is also available.
+//
+// A TypedCustom is created by calling [BindCustomResource] on a [Custom] with a
+// Go type to which it is assignable (see [thema.BindType]).
+type TypedCustom[R Resource] interface {
+	Custom
+
+	// TypeFromBytes is the same as [Custom.FromBytes], but returns an instance of the
+	// associated generic struct type instead of an [UnstructuredResource].
+	TypeFromBytes(b []byte, codec Decoder) (R, error)
```

Comment:
Same comment about the explicit `Decoder` vs accepting `encoding.GrafanaShapeBytes`